trace/workflow_state_worker.go

```diff
@@ -0,0 +1,139 @@
+	if description, err := client.DescribeWorkflowExecution(ctx, state.Execution.GetWorkflowId(), state.Execution.GetRunId()); err == nil {
+		execInfo := description.GetWorkflowExecutionInfo()
+		state.HistoryLength = execInfo.HistoryLength
+		state.IsArchived = execInfo.HistoryLength == 0 // TODO: Find a better way to identify archived workflows
+	}
+
+	return &WorkflowStateJob{
+		ctx:        ctx,
+		client:     client,
+		state:      state,
+		depth:      depth,
+		fetchAll:   fetchAll,
+		updateChan: updateChan,
+		childJobs:  []*WorkflowStateJob{},
+	}, nil
+}
+
+// Run starts the WorkflowStateJob, which retrieves the workflow's events and spawns new jobs for the child workflows once it's up-to-date.
+// New jobs are submitted to the pool when the job is up-to-date to reduce the amount of unnecessary history fetches (e.g. when the child workflow is already completed).
+func (job *WorkflowStateJob) Run(group *pond.TaskGroupWithContext) func() error {
+	return func() error {
+		state := job.state
+		wfId := state.Execution.GetWorkflowId()
+		runId := state.Execution.GetRunId()
+
+		// Make sure to not long poll archived workflows since GetWorkflowHistory fails under those circumstances.
+		isLongPoll := !state.IsArchived
+		historyIterator := job.client.GetWorkflowHistory(job.ctx, wfId, runId, isLongPoll, enums.HISTORY_EVENT_FILTER_TYPE_ALL_EVENT)
+
+		for historyIterator.HasNext() {
+			event, err := historyIterator.Next()
+			if err != nil {
+				return err
+			}
+			if event == nil {
+				continue
+			}
+
+			// Update state with new event and signal
+			job.state.Update(event)
+			job.updateChan <- struct{}{}
+
+			// Create child jobs if we're on a child workflow execution started event and we haven't hit depth 0.
+			if event.EventType == enums.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_STARTED && job.depth != 0 {
+				childJob, err := job.GetChildJob(event)
+				if err != nil {
+					// TODO: Consider if we want to error out if a child workflow cannot be updated by itself.
+					return err
+				}
+				job.childJobs = append(job.childJobs, childJob)
+
+				if job.isUpToDate {
+					group.Submit(childJob.Run(group))
+				}
+			}
+
+			// Start child jobs when we're up-to-date
+			if !job.isUpToDate && event.EventId >= state.HistoryLength {
+				job.isUpToDate = true
+				for _, childJob := range job.childJobs {
+					if childJob.ShouldStart() {
+						group.Submit(childJob.Run(group))
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+}
+
+// GetChildJob gets a new child job and appends it to the list of childJobs. These jobs don't start until the parent is catched up.
+func (job *WorkflowStateJob) GetChildJob(event *history.HistoryEvent) (*WorkflowStateJob, error) {
+	// Retrieve child workflow from parent and create a new job to fetch events for it
+	childAttrs := event.GetChildWorkflowExecutionStartedEventAttributes()
+	wf, ok := job.state.GetChildWorkflowByEventId(childAttrs.GetInitiatedEventId())
```

Comment:
Looks like `wf` is accessed concurrently, so I'm now convinced we should have a lock for the entire struct.

Reply:
That's what the lock in `GetChildWorkflowByEventId` is for.

Reply:
Yes but then you pass `wf` to `NewWorkflowStateJob` unlocked.

Again this might not matter today since we update a couple of attributes (`HistoryLength` and `IsArchived`) that may not need to be atomically updated but, as mentioned above, I'd rather be on the safe side and make sure all access to this struct is serialized.

Reply:
Each state struct would have its own lock, you wouldn't hold the same lock for all workflows.
Again, I'd rather have code that makes less assumptions about concurrent usage but I think we can move on to the next PR.